File: templates/problems/04_stack/07_largest_rect_histogram/typescript/solution.ts
/**
 * Use a monotonic increasing stack.
 * Time: O(n), Space: O(n)
 */
export function solveLargestRectangleArea(heights: number[]): number {
  const stack: number[] = []; // stack of indices
  let maxArea = 0;
  for (let i = 0; i <= heights.length; i++) {
    const h = i < heights.length ? heights[i] : 0;
    while (stack.length > 0 && h < heights[stack[stack.length - 1]]) {
      const height = heights[stack.pop()!];
      const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
      const area = height * width;
      if (area > maxArea) {
        maxArea = area;
      }
    }
    stack.push(i);
  }
  return maxArea;
}

File: templates/problems/04_stack/07_largest_rect_histogram/python/solution.py
def solve_largest_rectangle_area(heights: list[int]) -> int:
    """Use a monotonic increasing stack.

    Time: O(n), Space: O(n)
    """
    stack: list[int] = []  # stack of indices
    max_area = 0
    for i in range(len(heights) + 1):
        h = heights[i] if i < len(heights) else 0
        while stack and h < heights[stack[-1]]:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            area = height * width
            if area > max_area:
                max_area = area
        stack.append(i)
    return max_area
